### ISSUE TITLE
lumberjack plugin pipeline blocked/closed SizedQueueTimeout

### ISSUE THREAD

**Issue by christofhaerens** [elastic/logstash#4447]
Lumberjack plugin always blocks after some time (4-15 minues)
We have 500 clients sending in around 8000 events/minute via logstash forwarder 0.4.0 (LSF)
The setup was:
LSF clients (x500) -> logstash server (x2) v1.5.4 -> ES (5 nodes) v1.7

The logstash servers have quite some grokking filters. Since we got warnings and error about pipeline slowdown and close we decided to split the filter workers from the incoming events.

So setup is now:
LSF clients (x500) -> logstash 'broker' (x2) -> redis (x2) -> logstash server (x2) -> ES (x5)

But here we experience the same problems. The LS brokers have no filters. We have tested with logstash 1.5.4 and 2.1.1 (currently on 2.1.1). But the problem isthe same with both versions. Logstash brokers have 8GB mem (of which 70% is configured as LS_HEAP_SIZE) and 4 cpu. Open files in logstash is set to 32k and to 65k on redhat 7 linux. nproc is set to 32k.
When we restart the LS-brokers they start taking the LSF events back in. The redis queue can even reach more then 1.000.000 events while the LSF clients are emptying their queue. Mem/cpu during that period on the LS-brokers is fine. After some time (no pattern yet detected) the slowdown starts and eventually close of the lumberjack plugin. Java proc is still running when the plugin is closed and no errors or warnings are logged.. Threads can go upto 1700 and then down to around 30.
When we scan for connections to the redis we see that the broker is shipping events to redis and sometimes re-opens the lumberjack. But eventually the lumberjack is closed again and never reopens. Even when the connections to the redis stop. All this happens within 4-15 minutes after the LS-brokers are started. Only way to get things running again is restarting logstash brokers. But restarting them every 15 min (or less) is not really an option.
We are considering filebeat but we can't of course switch 500 clients from LSF to filebeat at this moment.

Already have set the lumberjack workers to 3 but then we get errors about multiline in the logstash logfile, despite we don't use the multiline filter (no filters are used at all).

Any help/ideas are welcome. If I can run it with some special options to produce more logging about why the plugin is closed and not re-opening again, I'm happy to try then out.

Cheers
Christof

**Comment by christofhaerens** [comment]
Here you can see how often and which events occured the logstash.log:
      6 =>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
   2386 =>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
  15484 =>"CircuitBreaker::rescuing exceptions", :name=>"Lumberjack input", :exception=>LogStash::SizedQueueTimeout::TimeoutError, :level=>:warn}
  15484 =>"Lumberjack input: The circuit breaker has detected a slowdown or stall in the pipeline, the input is closing the current connection and rejecting new connection until the pipeline recover.", :exception=>LogStash::CircuitBreaker::HalfOpenBreaker, :level=>:warn}
   2386 =>"Lumberjack input: The circuit breaker has detected a slowdown or stall in the pipeline, the input is closing the current connection and rejecting new connection until the pipeline recover.", :exception=>LogStash::CircuitBreaker::OpenBreaker, :level=>:warn}
    350 =>"Lumberjack input: the pipeline is blocked, temporary refusing new connection.", :level=>:warn}
      4 =>"SIGTERM received. Shutting down the pipeline.", :level=>:warn}

(aggr on the logfile since this morning)

**Comment by suyograo** [comment]
@ph is this related to https://github.com/logstash-plugins/logstash-input-beats/issues/37 ?

**Comment by ph** [comment]
@suyograo  I believe its not the same issue, since I don't see a specific `Concurrency` errors, @christofhaerens can you check if you can find any mention of it?

@christofhaerens Can you share the configuration of the broker?  I know its minimal but could always help!

I have a few questions:
- There is no redis related error at all?
- Nothing mention that the breaker is trying to reopen? 
- Are you using the the `workers` attribute for the redis output?

I understand It can be hard to update to filebeat, but the new light client has more logic to recover from a contention scenario and the beats input has quite a bit more logging now.

**Comment by ph** [comment]
No mention of `Failed to send event to Redis` ? What I think could help here is to set the `workers` attribute of the redis output to something higher, start with 4-5, by default it will use a single thread to send events to redis.

Another thing you can play with the `congestion_threshold` option of the lumberjack input try settings to 10. This will tell the breaker to wait a bit more time before closing, the default is 5.

**Comment by christofhaerens** [comment]
Conf file:
```
input { 
  lumberjack {
    port => 6782
    ssl_certificate => "/etc/puppetlabs/puppet/ssl/certs/xxx.pem"
    ssl_key => "/etc/puppetlabs/puppet/ssl/private_keys/xxx.pem"
    type => "lumberjack"
  }
}
input {
  udp {
    port        => 2515
    type        => "firewall"
  }
}
input {
  udp {
    port        => 2514
    type        => "isam"
  }
}
input {
  udp {
    port        => 2516
    type        => "loadbalancer"
  }
}
output {
  redis {
    host => ["elk-redis1", "elk-redis2"]
    codec => "json"
    data_type => "list"
    key => "logstash"
   #workers => 3
  }
}
```

Redis is not under stress at all or produces any errors.
In my second post to this issue I have stripped the log file of the timestamp and did then a sort and count on it. So that are all the different messages that are in the file. No messages of reopen, no messages that redis failed. So no mention of Concurrency errors
As you can see from the conf file, I tried with the workers in the redis output. Which lead to errors about the multiline filter. Have looked also at the ```congestion_threshold``` of the input plugin but haven't played with it yet since the doc was not 100% clear on what it did. Will try setting it at 10

java version
```
# java -version
java version "1.7.0_79"
OpenJDK Runtime Environment (novell-2.5.5.1.el7_1-x86_64 u79-b14)
OpenJDK 64-Bit Server VM (build 24.79-b02, mixed mode)
```

I can install a jre 1.8 if needed.

**Comment by christofhaerens** [comment]
forgot to mention. Udp listeners are defined but not used yet. The now still go to the logstash servers directly. Once the broker is stable we will also move udp traffic to the broker.

**Comment by christofhaerens** [comment]
Tomorrow (already 21h34 here) I will do a little poc of filebeat and see how easy we can get our client module in puppet changed to filebeat. Once that is done, commited and tested I can change 500 clients in a matter of minutes :-)

**Comment by ph** [comment]
@christofhaerens 

From your broker configuration I don't see any mention of multiline codec OR filter, so I think the error you had was with the previous configuration that was doing all the work. Also this error was probably due of using the `-w` option when starting logstash.

From what you are saying redis isn't under pressure at all, so a bit of background about the way logstash deal with events.

- First using `workers` in the lumberjack input wont do anything, since each remove server connection create his own thread.
- Logstash has a minimal internal queue, currently set at 20 items in the version you are using, this number is low to reduce the data loss possibility.
- Events are popped from the queue and send to the redis output

With your current configuration, a single thread is pushing the event one at a time to the redis server. Which might not be optimal under load, the required IO could reduce the performance of removing items from the queue and might cause the breaker excepton you had.

So with your configuration and settings redis workers set to 4 in the redis output, it should give you 4 thread to do the work.

```
  redis {
    host => ["elk-redis1", "elk-redis2"]
    codec => "json"
    data_type => "list"
    key => "logstash"
    workers => 4
  }
}
```

Also this wont fix the problem of the output sending one item at a time, but the output support a batch option, the output will buffer the event and send it in batch to redis. This operation is usually faster than  sending an event at a time. (https://github.com/logstash-plugins/logstash-output-redis/blob/master/lib/logstash/outputs/redis.rb#L73)

```
  redis {
    host => ["elk-redis1", "elk-redis2"]
    codec => "json"
    data_type => "list"
    key => "logstash"
    workers => 4
    batch => true
  }
}
```

This will mean, you have 4 workers buffering 50 (the default) before pushing this to redis.

Also in your case the `-w` option wont change anything since you don't have any filters. (this will change in 2.2)

**Comment by christofhaerens** [comment]
congestion_threshold set to 10 and restarted. 1 broker pipeline closed. Other node still running but prob not for long since all clients now will shift to that node.

**Comment by ph** [comment]
Also you don't see this 'CircuitBreaker::Close' in your log?

**Comment by christofhaerens** [comment]
@ph 
yes I see the close:
```
# grep "CircuitBreaker::Close" /var/log/logstash/logstash.log
{:timestamp=>"2016-01-11T09:45:40.207000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T10:00:37.236000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T10:30:38.562000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T10:45:37.075000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T14:00:39.183000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T14:23:43.321000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T14:44:05.779000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T15:15:38.839000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T15:45:38.124000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T16:00:36.844000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T16:22:43.063000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:54.901000+0100", :message=>"CircuitBreaker::Close", :name=>"Lumberjack input", :level=>:warn}
````

I'm sure the multiline error was on the broker. (Besides we don't the use the multiline filter at all, since we have our tomcats logging in json straight away)

I will set the ```workers``` again and also set the ```batch``` option and see how that goes.

**Comment by christofhaerens** [comment]
@ph If you meantt with re-open the open message: Yes we get those just up to a moment it blocks and doesn't open again:

```
# grep '"CircuitBreaker::Open"' /var/log/logstash/logstash.log |wc -l
10323
# grep '"CircuitBreaker::Open"' /var/log/logstash/logstash.log |tail
{:timestamp=>"2016-01-11T21:45:18.417000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.419000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.421000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.421000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.421000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.422000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.423000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.424000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.425000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
{:timestamp=>"2016-01-11T21:45:18.426000+0100", :message=>"CircuitBreaker::Open", :name=>"Lumberjack input", :level=>:warn}
```

**Comment by ph** [comment]
@christofhaerens Yeah, the breaker is trying to close to let the events flow again, So I believe the input is fine since he is trying to "recover" and this make me suspect the slow down is in the output. This doesn't mean the redis server is the problem.

We will find the issues :)

The batch is also something we use with the elasticsearch output, by default the output is buffering 500 events before actually using the bulk api.

Concerning Filebeat, when a contention happen, the same scenario occur, but FB will do a slow start to test if the logstash server is healthy before sending all the data.

**Comment by ph** [comment]
I haven't seen the case it never reopen, so this might be another issue.

**Comment by christofhaerens** [comment]
OK little update before I signoff for today:
both brokers are doing nice so far, not a single slowdown after 12 min running. They are filling up both redis with already 1M events in each redis.
But my events no longer show up in the elasticsearch. Guess I will have to set the ```batch_count``` to 50 on the redis input of the logstash servers.

**Comment by ph** [comment]
@christofhaerens yes they need to match! 
Ping me tomorrow if that solve your issue! :)

**Comment by christofhaerens** [comment]
@ph seems the load can be handled now. Java mem footprint now also much lower. Not a single slowdown since the LS-brokers were restarted this night for the log rotation.
Thxs for the help and the explanation of the logstash internals.

**Comment by ph** [comment]
@christofhaerens glad it worked!

**Comment by christofhaerens** [comment]
@ph
Even added some extra logs and now we are up to 25-30k events per minute. LS-broker mem ~ 500MB.
Quite a change with before!

**Comment by ph** [comment]
@christofhaerens Also It might be possible to use more workers than 4 for the redis output, so If you are experiencing the same problems try to increment theses values before adding new broker.

Since the workers are waiting on IO/connection on redis you can have the benefit to run more.

**Comment by christofhaerens** [comment]
@ph idd because I also monitor which redis the event passes. Since the redis output takes one out the host list, it was often that both brokers picked the same and that only 1 redis was used. Now the distribution of redis usage is around 50% each. So each worker is rotating through the host list, leveling out the events over both redis machines.
